addons/l10n_tw_edi_ecpay/tests/test_edi.py

```diff
@@ -0,0 +1,338 @@
+            'company_type': 'person',
+        })
+
+        # We can reuse this invoice for the flow tests.
+        cls.basic_invoice = cls.init_invoice(
+            'out_invoice', partner=cls.partner_a, products=cls.product_a,
+        )
+        cls.basic_invoice.action_post()
+
+    def test_01_can_generate_file(self):
+        """
+        Simply test that with a valid configuration, we can generate the file.
+        """
+        with patch(CALL_API_METHOD, new=self._test_01_mock):
+            json_data = self.basic_invoice._l10n_tw_edi_generate_invoice_json()
+        self.assertTrue(json_data)
+
+        formatted_phone = self.basic_invoice._reformat_phone_number(self.basic_invoice.partner_id.phone)
+        # Validate the customer data
+        self.assertEqual(json_data.get("MerchantID"), self.company_data['company'].l10n_tw_edi_ecpay_merchant_id)
+        self.assertEqual(json_data.get("CustomerName"), self.basic_invoice.commercial_partner_id.name)
+        self.assertEqual(json_data.get("CustomerAddr"), self.basic_invoice.partner_id.contact_address)
+        self.assertEqual(json_data.get("CustomerEmail"), self.basic_invoice.partner_id.email)
+        self.assertEqual(json_data.get("CustomerPhone"), formatted_phone)
+        self.assertEqual(json_data.get("SalesAmount"), self.basic_invoice.amount_total)
+
+    @freeze_time("2025-01-06 15:00:00")
+    def test_02_basic_submission(self):
+        """
+        This tests the most basic flow: an invoice is successfully sent to the ECpay platform, and then pass validation.
+        """
+        send_and_print = self.create_send_and_print(self.basic_invoice)
+        with patch(CALL_API_METHOD, new=self._test_02_mock):
+            send_and_print.action_send_and_print()
+
+        # Now that the invoice has been sent successfully, we assert that some info have been saved correctly.
+        self.assertRecordValues(
+            self.basic_invoice,
+            [{
+                'l10n_tw_edi_ecpay_invoice_id': 'AB11100099',
+                'l10n_tw_edi_invoice_create_date': datetime.strptime('2025-01-06 15:00:00', '%Y-%m-%d %H:%M:%S'),
+                'l10n_tw_edi_state': 'valid',
+            }]
+        )
+
+        self.assertTrue(self.basic_invoice.l10n_tw_edi_file_id)
+
+    def test_03_failed_submission(self):
+        """
+        This test will test a flow where the submission itself (not the documents inside) fails for any reason.
+        A general error as such should be handled, but is not expected and should be treated as a bug on our side.
+
+        As we submit a single invoice, we expect a UserError to be raised.
+        """
+        send_and_print = self.create_send_and_print(self.basic_invoice)
+        with patch(CALL_API_METHOD, new=self._test_03_mock):
+            with self.assertRaises(UserError):
+                send_and_print.action_send_and_print()
+
+    @freeze_time("2025-01-06 15:00:00")
+    def test_04_invalid_invoice(self):
```

Comment:
In general, it's good practice to give a short description of what each test aims to do as docstring.